cmd/mount_zfs/mount_zfs.c

```diff
@@ -47,63 +47,43 @@ libzfs_handle_t *g_zfs;
  * is expected to be of the form pool/dataset, however may also refer to
  * a block device if that device contains a valid zfs label.
  */
-static char *
-parse_dataset(char *dataset)
+static void
+parse_dataset(char (*dataset)[PATH_MAX])
 {
-	char cwd[PATH_MAX];
-	struct stat64 statbuf;
-	int error;
-	int len;
-
 	/*
 	 * We expect a pool/dataset to be provided, however if we're
 	 * given a device which is a member of a zpool we attempt to
 	 * extract the pool name stored in the label.  Given the pool
 	 * name we can mount the root dataset.
 	 */
-	error = stat64(dataset, &statbuf);
-	if (error == 0) {
+	int fd = open((const char *)dataset, O_RDONLY);
+	if (fd > 0) {
 		nvlist_t *config;
-		char *name;
-		int fd;
-
-		fd = open(dataset, O_RDONLY);
-		if (fd < 0)
-			goto out;
-
-		error = zpool_read_label(fd, &config, NULL);
+		char *name = NULL;
+		zpool_read_label(fd, &config, NULL);
 		(void) close(fd);
-		if (error)
-			goto out;
-
-		error = nvlist_lookup_string(config,
-		    ZPOOL_CONFIG_POOL_NAME, &name);
-		if (error) {
-			nvlist_free(config);
-		} else {
-			dataset = strdup(name);
-			nvlist_free(config);
-			return (dataset);
-		}
+
+		nvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME, &name);
```

Comment:
Here too.

Reply:
`nvlist_lookup_string` (and `nvlist_free`) bails early on a NULL `nvlist_t*`, and on failure can leave `name` in only two potential states: the untouched NULL it was initialized with, or a freshly assigned NULL. So we can safely fall through to the single return statement.

Reply:
How about an alternate solution like this when instead depends on the authoritative return values and makes clear when `config` and `name` are NULL.

```c
                error = zpool_read_label(fd, &config, NULL);
                (void) close(fd);
                if (error == 0) {
                        error = nvlist_lookup_string(config,
                            ZPOOL_CONFIG_POOL_NAME, &name);
                        if (error == 0)
                                (void) strlcpy(*dataset, name, PATH_MAX);

                        nvlist_free(config);
                }
```